src/GrantRegistrar.sol

```diff
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import { Ownable } from 'openzeppelin-contracts/contracts/access/Ownable.sol';
+import { IWorldIDGroups } from 'world-id-contracts/interfaces/IWorldIDGroups.sol';
+import { Ownable2Step } from 'openzeppelin-contracts/contracts/access/Ownable2Step.sol';
+
+/// @title GrantRegistrar
+/// @author Worldcoin
+contract GrantRegistrar is Ownable2Step {
+  ///////////////////////////////////////////////////////////////////////////////
+  ///                              CONFIG STORAGE                            ///
+  //////////////////////////////////////////////////////////////////////////////
+
+  /// @dev The WorldID router instance that will be used for managing groups and verifying proofs
+  IWorldIDGroups public worldIdRouter;
+
+  /// @dev The World ID group whose participants can verify with this contract.
+  uint256 public groupId;
+
+  /// @dev The World ID nullifier hash that will be used to verify proofs.
+  uint256 immutable externalNullifierHash;
+
+  /// @dev The maximum grant ID that can be claimed without re-verifying.
+  uint256 public maxGrantId;
```

Comment:
We should calculate this dynamically, updating this regularly will be very annoying and error prone.
Since we now bind it to grantIds, we can do something like this:
```solidity
maxGrantId = (((grant.getCurrentId()-offset) / nGrantsValidity) + 1) * nGrantsValidity - 1;
```
We can set `offset=grant.getCurrentId()` once in the constructor.
We should use the same grant as in the `RecurringGrantDrop` (maybe even worth calling the getter?). `nGrantsValidity` should be dynamically changeable (we can never really decrease it though). Then we know that verifications are always valid for `nGrantsValidity` grants. To actually have some benefits of this whole exercise, I would suggest something like `nGrantsValidity=6`. That grants may change in time doesn't matter too much as long as the duration of those `nGrantsValidity` grants is smaller than the deletion window. So that's something we need to watch out for when extending grants in the future. With the current deletion windows this is not really a concern.